File: frontend/src/types/listing.ts
export interface Enrichment {
  avg_rent_area: number | null
  population_density: number | null
  commercial_density: number | null
  transport_score: number | null
  liquidity_score: number | null
  accessibility_score: number | null
  vertical_storage_potential: number | null
  price_per_sqm: number | null
  estimated_rent_low: number | null
  estimated_rent_high: number | null
  gross_yield: number | null
  storage_yield_estimate: number | null
  ml_estimated_price: number | null
  ml_price_deviation: number | null
  edge_score: number | null
  computed_at: string | null
}

export interface Listing {
  id: string
  source: string
  external_id: string | null
  url: string | null
  title: string
  description: string | null
  price: number
  surface: number | null
  city: string | null
  postal_code: string | null
  address: string | null
  photos_count: number
  floor_level: number | null
  accessibility_tags: string[]
  lat: number | null
  lon: number | null
  scraped_at: string
  updated_at: string
  enrichment: Enrichment | null
}

export interface ListingsPage {
  total: number
  items: Listing[]
}

export interface GeoFeatureProperties {
  id: string
  title: string
  price: number
  surface: number | null
  city: string | null
  url: string | null
  edge_score: number | null
  gross_yield: number | null
  price_per_sqm: number | null
}

export interface GeoFeature {
  type: 'Feature'
  geometry: { type: 'Point'; coordinates: [number, number] }
  properties: GeoFeatureProperties
}

export interface GeoJSON {
  type: 'FeatureCollection'
  features: GeoFeature[]
}

export interface DashboardSummary {
  total_listings: number
  enriched_listings: number
  avg_edge_score: number | null
  avg_gross_yield: number | null
  avg_price: number | null
  top_cities: { city: string; count: number }[]
}

export interface TopOpportunity {
  id: string
  title: string
  city: string | null
  price: number
  surface: number | null
  gross_yield: number | null
  edge_score: number
  url: string | null
}

export interface Filters {
  city: string
  minPrice: number | null
  maxPrice: number | null
  minSurface: number | null
  maxSurface: number | null
  minYield: number | null
  minEdge: number | null
}

File: backend/app/schemas/listing.py
import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field


class EnrichmentSchema(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    avg_rent_area: float | None = None
    population_density: float | None = None
    commercial_density: float | None = None
    transport_score: float | None = None
    liquidity_score: float | None = None
    accessibility_score: float | None = None
    vertical_storage_potential: float | None = None
    price_per_sqm: float | None = None
    estimated_rent_low: float | None = None
    estimated_rent_high: float | None = None
    gross_yield: float | None = None
    storage_yield_estimate: float | None = None
    ml_estimated_price: float | None = None
    ml_price_deviation: float | None = None
    edge_score: float | None = None
    computed_at: datetime | None = None


class ListingBase(BaseModel):
    title: str
    source: str
    price: float
    surface: float | None = None
    city: str | None = None
    postal_code: str | None = None
    address: str | None = None
    photos_count: int = 0
    floor_level: int | None = None
    accessibility_tags: list[str] = Field(default_factory=list)
    url: str | None = None
    description: str | None = None


class ListingCreate(ListingBase):
    external_id: str | None = None
    lat: float | None = None
    lon: float | None = None


class ListingSchema(ListingBase):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    external_id: str | None = None
    lat: float | None = None
    lon: float | None = None
    scraped_at: datetime
    updated_at: datetime
    enrichment: EnrichmentSchema | None = None


class ListingGeoFeature(BaseModel):
    type: str = "Feature"
    geometry: dict
    properties: dict


class ListingGeoJSON(BaseModel):
    type: str = "FeatureCollection"
    features: list[ListingGeoFeature]


class ListingsPage(BaseModel):
    total: int
    items: list[ListingSchema]
